File: chrome/android/java/src/org/chromium/chrome/browser/autofill/AutofillExternalDelegate.java
// Copyright (c) 2012 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.autofill;

import android.graphics.Rect;

import org.chromium.base.CalledByNative;
import org.chromium.content.browser.ContainerViewDelegate;
import org.chromium.ui.gfx.NativeWindow;

/**
 * Android specific external delegate for display and selection of autofill data.
 */
public class AutofillExternalDelegate {

    // Area of the Autofill popup anchor position used to position the Autofill window.
    private Rect mAutofillAnchorRect;
    private AutofillWindow mAutofillWindow;
    private final int mNativeAutofillExternalDelegate;

    /**
     * Creates an AutofillExternalDelegate object.
     * @param nativeAutofillExternalDelegate A pointer to the native AutofillExternalDelegate.
     */
    public AutofillExternalDelegate(int nativeAutofillExternalDelegate) {
        mNativeAutofillExternalDelegate = nativeAutofillExternalDelegate;
    }

    @CalledByNative
    private static AutofillExternalDelegate create(int nativeAutofillExternalDelegate) {
        return new AutofillExternalDelegate(nativeAutofillExternalDelegate);
    }

    /**
     * Informs the native delegate that an Autofill suggestion has been chosen.
     * @param listIndex The index of the autofill suggestion selected
     * @param value The value of the autofill suggestion.
     * @param uniqueId The unique ID of the autofill suggestion selected.
     */
    public void autofillPopupSelected(int listIndex, String value, int uniqueId) {
        if (mNativeAutofillExternalDelegate != 0) {
            nativeAutofillPopupSelected(
                    mNativeAutofillExternalDelegate, listIndex, value, uniqueId);
        }
    }

    @SuppressWarnings("unused")
    @CalledByNative
    private void openAutofillPopup(ContainerViewDelegate containerViewDelegate,
            NativeWindow nativeWindow, AutofillSuggestion[] data) {
        if (mAutofillWindow == null) {
            mAutofillWindow = new AutofillWindow(nativeWindow, containerViewDelegate, this, data);
        } else {
            mAutofillWindow.setAutofillSuggestions(data);
        }
        assert mAutofillAnchorRect != null :
                "Autofill anchor rect should have been set by setBounds()";
        mAutofillWindow.setPositionAround(mAutofillAnchorRect);
        mAutofillWindow.show();
    }

    @SuppressWarnings("unused")
    @CalledByNative
    private void closeAutofillPopup() {
        if (mAutofillWindow == null) return;
        mAutofillWindow.dismiss();
        mAutofillWindow = null;
        mAutofillAnchorRect = null;
    }

    @SuppressWarnings("unused")
    @CalledByNative
    private void setAutofillAnchorRect(float x, float y, float width, float height) {
        mAutofillAnchorRect = new Rect((int) x, (int) y, (int) (x + width), (int) (y + height));
    }

    @SuppressWarnings("unused")
    @CalledByNative
    private static AutofillSuggestion createAutofillSuggestion(String name, String label,
            int uniqueId) {
        return new AutofillSuggestion(name, label, uniqueId);
    }

    private native void nativeAutofillPopupSelected(int nativeAutofillExternalDelegateAndroid,
            int listIndex, String value, int uniqueId);
}

File: chrome/browser/ui/android/autofill/autofill_external_delegate.cc
// Copyright (c) 2012 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/browser/ui/android/autofill/autofill_external_delegate.h"

#include "base/string_util.h"
#include "chrome/browser/autofill/autofill_manager.h"
#include "chrome/browser/ui/android/autofill/autofill_popup_view_android.h"
#include "content/public/browser/render_view_host.h"
#include "content/public/browser/web_contents.h"

void AutofillExternalDelegate::CreateForWebContentsAndManager(
    content::WebContents* web_contents,
    AutofillManager* autofill_manager) {
  if (FromWebContents(web_contents))
    return;

  web_contents->SetUserData(
      UserDataKey(),
      new AutofillExternalDelegateAndroid(web_contents, autofill_manager));
}

AutofillExternalDelegateAndroid::AutofillExternalDelegateAndroid(
    content::WebContents* web_contents, AutofillManager* manager)
    : AutofillExternalDelegate(web_contents, manager) {
}

AutofillExternalDelegateAndroid::~AutofillExternalDelegateAndroid() {
}

void AutofillExternalDelegateAndroid::ApplyAutofillSuggestions(
    const std::vector<string16>& autofill_values,
    const std::vector<string16>& autofill_labels,
    const std::vector<string16>& autofill_icons,
    const std::vector<int>& autofill_unique_ids) {
  view_->Show(autofill_values,
              autofill_labels,
              autofill_icons,
              autofill_unique_ids);
}

void AutofillExternalDelegateAndroid::HideAutofillPopupInternal() {
  if (!view_.get())
    return;

  view_->Hide();
  view_.reset();
}

void AutofillExternalDelegateAndroid::CreatePopupForElement(
    const gfx::Rect& element_bounds) {
  DCHECK(!view_.get());

  content::WebContents* contents = web_contents();
  view_.reset(new AutofillPopupViewAndroid(contents, this, element_bounds));
}
